Sampling from truncated mixture model results in an error

Hi! I have Julia 1.7.3 and Distributions 0.25.45. The error is as follows:

~~~~
julia> dist = Truncated(MixtureModel(Normal, [(1.0, 0.0)]), 0, Inf);

julia> rand(dist, 10)
ERROR: AssertionError: cl <= p <= cr
Stacktrace:
  [1] quantile_bisect(d::MixtureModel{Univariate, Continuous, Normal, Categorical{Float64, Vector{Float64}}}, p::Float64, lx::Float64, rx::Float64, tol::Float64)
    @ Distributions ~/.julia/packages/Distributions/9Albf/src/quantilealgs.jl:13
  [2] quantile_bisect(d::MixtureModel{Univariate, Continuous, Normal, Categorical{Float64, Vector{Float64}}}, p::Float64, lx::Float64, rx::Float64)
    @ Distributions ~/.julia/packages/Distributions/9Albf/src/quantilealgs.jl:11
  [3] quantile(d::MixtureModel{Univariate, Continuous, Normal, Categorical{Float64, Vector{Float64}}}, p::Float64)
    @ Distributions ~/.julia/packages/Distributions/9Albf/src/mixtures/mixturemodel.jl:447
  [4] quantile
    @ ~/.julia/packages/Distributions/9Albf/src/truncate.jl:128 [inlined]
  [5] rand
    @ ~/.julia/packages/Distributions/9Albf/src/univariates.jl:157 [inlined]
  [6] _rand!(rng::Random._GLOBAL_RNG, sampler::Truncated{MixtureModel{Univariate, Continuous, Normal, Categorical{Float64, Vector{Float64}}}, Continuous, Float64}, A::Vector{Float64})
    @ Distributions ~/.julia/packages/Distributions/9Albf/src/univariates.jl:147
  [7] rand!
    @ ~/.julia/packages/Distributions/9Albf/src/univariates.jl:142 [inlined]
  [8] rand
    @ ~/.julia/packages/Distributions/9Albf/src/genericrand.jl:52 [inlined]
  [9] rand
    @ ~/.julia/packages/Distributions/9Albf/src/genericrand.jl:24 [inlined]
 [10] rand(s::Truncated{MixtureModel{Univariate, Continuous, Normal, Categorical{Float64, Vector{Float64}}}, Continuous, Float64}, dims::Int64)
    @ Distributions ~/.julia/packages/Distributions/9Albf/src/genericrand.jl:22
 [11] top-level scope
    @ REPL[21]:1
~~~~

--- a/Project.toml
+++ b/Project.toml
@@ -1,7 +1,7 @@
 name = "Distributions"
 uuid = "31c24e10-a181-5473-b8eb-7969acd0382f"
 authors = ["JuliaStats"]
-version = "0.25.47"
+version = "0.25.48"
 
 [deps]
 ChainRulesCore = "d360d2e6-b24c-11e9-a2a3-2a2ae2dbcce4"

--- a/Project.toml
+++ b/Project.toml
@@ -1,7 +1,7 @@
 name = "Distributions"
 uuid = "31c24e10-a181-5473-b8eb-7969acd0382f"
 authors = ["JuliaStats"]
-version = "0.25.46"
+version = "0.25.47"
 
 [deps]
 ChainRulesCore = "d360d2e6-b24c-11e9-a2a3-2a2ae2dbcce4"

--- a/src/quantilealgs.jl
+++ b/src/quantilealgs.jl
@@ -15,7 +15,6 @@ function quantile_bisect(d::ContinuousUnivariateDistribution, p::Real, lx::T, rx
     # ≈ 3.7e-11 for Float64
     # ≈ 2.4e-5 for Float32
     tol = eps(float(T))^(2 / 3)
-    
     # find quantile using bisect algorithm
     cl = cdf(d, lx)
     cr = cdf(d, rx)

--- a/src/quantilealgs.jl
+++ b/src/quantilealgs.jl
@@ -1,7 +1,7 @@
 # Various algorithms for computing quantile
 
 function quantile_bisect(d::ContinuousUnivariateDistribution, p::Real, lx::T, rx::T) where {T<:Real}
-    rx < lx && throw(ArgumentError("[$lx, $rx] is not a valid bracketing interval"))
+    rx < lx && throw(ArgumentError("empty bracketing interval [$lx, $rx]"))
 
     # In some special cases, e.g. #1501, rx == lx`
     # If the distribution is degenerate the check below can fail, hence we skip it
@@ -14,7 +14,7 @@ function quantile_bisect(d::ContinuousUnivariateDistribution, p::Real, lx::T, rx
     # base tolerance on types to support e.g. `Float32` (avoids an infinite loop)
     # ≈ 3.7e-11 for Float64
     # ≈ 2.4e-5 for Float32
-    tol = eps(float(T))^(2 / 3)
+    tol = cbrt(eps(float(T)))^2
     # find quantile using bisect algorithm
     cl = cdf(d, lx)
     cr = cdf(d, rx)

--- a/src/quantilealgs.jl
+++ b/src/quantilealgs.jl
@@ -1,16 +1,26 @@
 # Various algorithms for computing quantile
 
-function quantile_bisect(
-    d::ContinuousUnivariateDistribution, p::Real, lx::Real, rx::Real,
+function quantile_bisect(d::ContinuousUnivariateDistribution, p::Real, lx::T, rx::T) where {T<:Real}
+    rx < lx && throw(ArgumentError("[$lx, $rx] is not a valid bracketing interval"))
+
+    # In some special cases, e.g. #1501, rx == lx`
+    # If the distribution is degenerate the check below can fail, hence we skip it
+    if rx == lx
+        # Returns `lx` of the same type as `(lx + rx) / 2`
+        # For specific types such as `Float64` it is more performant than `oftype((lx + rx) / 2, lx)`
+        return middle(lx)
+    end
+
     # base tolerance on types to support e.g. `Float32` (avoids an infinite loop)
     # ≈ 3.7e-11 for Float64
     # ≈ 2.4e-5 for Float32
-    tol::Real=(eps(Base.promote_typeof(float(lx), float(rx))))^(2 / 3),
-)
+    tol = eps(float(T))^(2 / 3)
+    
     # find quantile using bisect algorithm
     cl = cdf(d, lx)
     cr = cdf(d, rx)
-    @assert cl <= p <= cr
+    cl <= p <= cr ||
+        throw(ArgumentError("[$lx, $rx] is not a valid bracketing interval for `quantile(d, $p)`"))
     while rx - lx > tol
         m = (lx + rx)/2
         c = cdf(d, m)
@@ -23,6 +33,10 @@ function quantile_bisect(
         end
     end
     return (lx + rx)/2
+end
+
+function quantile_bisect(d::ContinuousUnivariateDistribution, p::Real, lx::Real, rx::Real)
+    return quantile_bisect(d, p, promote(lx, rx)...)
 end
 
 quantile_bisect(d::ContinuousUnivariateDistribution, p::Real) =

--- a/test/mixture.jl
+++ b/test/mixture.jl
@@ -280,7 +280,7 @@ end
     # issue #1501
     @testset "quantile of mixture with single component" begin
         for T in (Float32, Float64)
-            d = MixtureModel(Normal{T}(T(1), T(0)))
+            d = MixtureModel([Normal{T}(T(1), T(0))])
             for p in (0.2, 0.2f0, 1//3)
                 @test @inferred(quantile(d, p)) == 1
                 @test @inferred(median(d)) == 1

--- a/test/mixture.jl
+++ b/test/mixture.jl
@@ -276,4 +276,15 @@ end
             @test cdf(utot,x) ≈ cdf(utot,x)
         end
     end
+
+    # issue #1501
+    @testset "quantile of mixture with single component" begin
+        for T in (Float32, Float64)
+            d = MixtureModel(Normal{T}(T(1), T(0)))
+            for p in (0.2, 0.2f0, 1//3)
+                @test @inferred(quantile(d, p)) == 1
+                @test @inferred(median(d)) == 1
+            end
+        end
+    end
 end

